File: src/data/tutorials/028-merge-k-sorted-lists.js
/**
 * Tutorial content for #028 — Merge K Sorted Lists
 * Keyed by approach name (must match the name parsed from the solution file).
 */
export const tutorial = {
  problem: {
    statement: `You are given an array of k linked lists, each sorted in ascending order. Merge all of them into one single sorted linked list and return its head.`,
    example: `Input: [[1→4→5], [1→3→4], [2→6]]\n→ Compare heads: 1, 1, 2 → pick 1\n→ Compare heads: 1, 3, 4, 2 → pick 1\n→ Continue picking smallest each time\n✅ Answer: 1 → 1 → 2 → 3 → 4 → 4 → 5 → 6`,
    keyInsight: `Merging k lists is expensive if you pick the minimum by scanning all k heads every time. A min-heap (priority queue) lets you find and remove the minimum in O(log k) instead of O(k), giving a much faster overall solution.`,
  },

  approaches: {
    'Min Heap': {
      intuition: `Put the head node of every list into a min-heap ordered by node value. Repeatedly extract the minimum node, append it to your result list, and if that node has a next node, push the next node into the heap. The heap always holds at most k nodes — one per remaining list — so each extraction is O(log k).`,
      steps: [
        `Create a min-heap (priority queue) and insert the head of each non-null list.`,
        `Create a dummy head node for the result list; keep a tail pointer on it.`,
        `While the heap is not empty: extract the node with the smallest value.`,
        `Append that node to tail (tail.next = node, then advance tail).`,
        `If the extracted node has a next pointer, push node.next into the heap.`,
        `Return dummy.next as the head of the merged list.`,
      ],
      example: `Lists: [1→4→5], [1→3→4], [2→6]\n\nHeap: {1(L1), 1(L2), 2(L3)}\nExtract 1(L1) → result: 1, push 4(L1). Heap: {1(L2),2(L3),4(L1)}\nExtract 1(L2) → result: 1→1, push 3(L2). Heap: {2(L3),3(L2),4(L1)}\nExtract 2(L3) → result: 1→1→2, push 6(L3). Heap: {3(L2),4(L1),6(L3)}\nExtract 3(L2) → result: 1→1→2→3, push 4(L2). Heap: {4(L1),4(L2),6(L3)}\n... continue until heap empty\n✅ Answer: 1→1→2→3→4→4→5→6`,
      keyInsight: `O(N log k) time where N = total nodes and k = number of lists. O(k) space for the heap. This is the optimal approach for large k.`,
    },

    'Divide and Conquer': {
      intuition: `Instead of merging all k lists at once, pair them up and merge each pair. Repeat this halving process — like a tournament bracket — until one list remains. Each round halves the number of lists, so there are log k rounds, and each round does O(N) work total across all merges.`,
      steps: [
        `Start with the array of k lists.`,
        `In each round, pair up adjacent lists and merge each pair using the classic two-list merge.`,
        `Collect the merged results as the new, smaller array of lists.`,
        `Repeat the pairing/merging until only one list remains.`,
        `Return that final list.`,
      ],
      example: `Lists: [L1, L2, L3, L4]  (k=4)\n\nRound 1: merge(L1,L2)=M12, merge(L3,L4)=M34 → [M12, M34]\nRound 2: merge(M12,M34)=M1234 → [M1234]\nDone!\n\nFor our example [[1→4→5],[1→3→4],[2→6]]:\nRound 1: merge(L1,L2)=[1→1→3→4→4→5], L3=[2→6] → 2 lists\nRound 2: merge → [1→1→2→3→4→4→5→6]\n✅ Answer: 1→1→2→3→4→4→5→6`,
      keyInsight: `O(N log k) time, O(1) extra space beyond recursion stack. Same time complexity as the heap but with very low constant factors. The divide-and-conquer structure avoids repeatedly re-merging long lists.`,
    },

    'Sequential Merge': {
      intuition: `Merge lists one by one: take the result so far and merge it with the next list in the array, then merge that result with the list after, and so on. You are reducing k lists to 1 by doing k-1 two-list merges sequentially.`,
      steps: [
        `If the input is empty, return null.`,
        `Set result = lists[0].`,
        `For i from 1 to k-1: result = mergeTwoLists(result, lists[i]).`,
        `Return result after all merges.`,
        `(mergeTwoLists uses the classic pointer-based O(n+m) merge.)`,
      ],
      example: `Lists: [1→4→5], [1→3→4], [2→6]\n\nStep 1: merge([1→4→5], [1→3→4]) = 1→1→3→4→4→5\nStep 2: merge([1→1→3→4→4→5], [2→6]) = 1→1→2→3→4→4→5→6\n✅ Answer: 1→1→2→3→4→4→5→6`,
      keyInsight: `O(kN) time in the worst case (the first list gets re-merged k-1 times, growing longer each round). O(1) extra space. Fine for small k but slow for large k — use heap or divide-and-conquer instead.`,
    },

    'Iterative Pairwise Merge': {
      intuition: `Merge lists two at a time in repeated rounds, similar to Divide and Conquer but implemented iteratively rather than recursively. In each round, pair up adjacent lists and merge each pair; the results become the input for the next round. After ⌈log k⌉ rounds a single list remains. This avoids the recursion stack of the divide-and-conquer approach.`,
      steps: [
        `Start with the array of k lists.`,
        `While the array has more than 1 list: create a new empty array for this round's results.`,
        `  Iterate i = 0, 2, 4, … over the current array: merge lists[i] and lists[i+1] (or just lists[i] if i+1 is out of bounds).`,
        `  Append each merged result to the round's result array.`,
        `Replace the current array with the round results and repeat.`,
        `Return the single remaining list.`,
      ],
      example: `Lists: [L1, L2, L3, L4, L5]  (k=5)\n\nRound 1 (pair adjacent):\n  merge(L1,L2)=M12, merge(L3,L4)=M34, L5 unpaired → [M12, M34, L5]\nRound 2:\n  merge(M12,M34)=M1234, L5 unpaired → [M1234, L5]\nRound 3:\n  merge(M1234,L5)=M12345 → [M12345]\nDone!\n\nFor [[1→4→5],[1→3→4],[2→6]]:\nRound 1: merge(L1,L2)=[1→1→3→4→4→5], L3=[2→6]\nRound 2: merge → [1→1→2→3→4→4→5→6]\n✅ Answer: 1→1→2→3→4→4→5→6`,
      keyInsight: `O(N log k) time, O(1) extra space (no recursion stack). The iterative pairing mirrors the divide-and-conquer structure but is easier to reason about stack depth — each round processes every node once and halves the list count.`,
    },

    'Collect + Sort': {
      intuition: `Drain all node values into a plain array, sort it with the language's built-in sort, then reconstruct a fresh linked list node by node. The name "Collect + Sort" distinguishes this from "Collect and Sort" (which builds new nodes); the two approaches are algorithmically identical but may differ in minor implementation details such as whether original nodes are reused.`,
      steps: [
        `Traverse every list and collect each node's value into a flat integer array.`,
        `Sort the array using the built-in comparator (e.g. Collections.sort or Arrays.sort).`,
        `Create a dummy head and a tail pointer for the result list.`,
        `Iterate over the sorted array: create a new ListNode for each value and append it to the tail.`,
        `Return dummy.next.`,
      ],
      example: `Lists: [1→4→5], [1→3→4], [2→6]\n\nCollect values: [1, 4, 5, 1, 3, 4, 2, 6]\nSort:           [1, 1, 2, 3, 4, 4, 5, 6]\nBuild new list: new node(1) → new node(1) → … → new node(6)\n✅ Answer: 1→1→2→3→4→4→5→6`,
      keyInsight: `O(N log N) time (sort dominates), O(N) space for the value array plus the new list. Simpler to implement than any merge-based strategy but sacrifices the sorted property of each input list.`,
    },
  },
}

File: public/cpp/028-merge-k-sorted-lists.cpp
/*
 * Problem: Merge K Sorted Lists (LeetCode 23)
 * Link: https://leetcode.com/problems/merge-k-sorted-lists/
 * Difficulty: Hard
 * Category: Linked Lists / Heaps
 * 
 * Description:
 * You are given an array of k linked-lists lists, each linked-list is sorted in
 * ascending order. Merge all the linked-lists into one sorted linked-list and return it.
 * 
 * Example 1:
 * Input: lists = [[1,4,5],[1,3,4],[2,6]]
 * Output: [1,1,2,3,4,4,5,6]
 * 
 * Example 2:
 * Input: lists = []
 * Output: []
 * 
 * Example 3:
 * Input: lists = [[]]
 * Output: []
 * 
 * Constraints:
 * - k == lists.length
 * - 0 <= k <= 10^4
 * - 0 <= lists[i].length <= 500
 * - -10^4 <= lists[i][j] <= 10^4
 * - lists[i] is sorted in ascending order.
 * - The sum of lists[i].length will not exceed 10^4.
 */

#include <iostream>
#include <vector>
#include <queue>
using namespace std;

// Definition for singly-linked list
struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

/*
 * APPROACH 1: MIN HEAP (OPTIMAL)
 * 
 * Intuition:
 * - Use min heap to track smallest element among k lists
 * - Always extract minimum and add its next node
 * - Heap size is at most k
 * 
 * Algorithm:
 * 1. Add first node of each list to min heap
 * 2. While heap not empty:
 *    - Extract minimum node
 *    - Add to result
 *    - If node has next, add next to heap
 * 3. Return merged list
 * 
 * Time Complexity: O(N log k) where N is total nodes, k is number of lists
 * Space Complexity: O(k) - heap size
 */
class Solution1 {
public:
    ListNode* mergeKLists(vector<ListNode*>& lists) {
        auto cmp = [](ListNode* a, ListNode* b) {
            return a->val > b->val;
        };
        priority_queue<ListNode*, vector<ListNode*>, decltype(cmp)> minHeap(cmp);
        
        // Add first node of each list
        for (ListNode* list : lists) {
            if (list) {
                minHeap.push(list);
            }
        }
        
        ListNode dummy(0);
        ListNode* tail = &dummy;
        
        while (!minHeap.empty()) {
            ListNode* node = minHeap.top();
            minHeap.pop();
            
            tail->next = node;
            tail = tail->next;
            
            if (node->next) {
                minHeap.push(node->next);
            }
        }
        
        return dummy.next;
    }
};

/*
 * APPROACH 2: DIVIDE AND CONQUER
 * 
 * Intuition:
 * - Merge lists pairwise
 * - Recursively merge pairs until one list remains
 * - Similar to merge sort
 * 
 * Algorithm:
 * 1. If 0 or 1 list, return it
 * 2. Divide lists into two halves
 * 3. Recursively merge each half
 * 4. Merge the two results
 * 
 * Time Complexity: O(N log k)
 * Space Complexity: O(log k) - recursion stack
 */
class Solution2 {
private:
    ListNode* mergeTwoLists(ListNode* l1, ListNode* l2) {
        ListNode dummy(0);
        ListNode* tail = &dummy;
        
        while (l1 && l2) {
            if (l1->val < l2->val) {
                tail->next = l1;
                l1 = l1->next;
            } else {
                tail->next = l2;
                l2 = l2->next;
            }
            tail = tail->next;
        }
        
        tail->next = l1 ? l1 : l2;
        return dummy.next;
    }
    
    ListNode* mergeKListsHelper(vector<ListNode*>& lists, int left, int right) {
        if (left > right) return nullptr;
        if (left == right) return lists[left];
        
        int mid = left + (right - left) / 2;
        ListNode* leftList = mergeKListsHelper(lists, left, mid);
        ListNode* rightList = mergeKListsHelper(lists, mid + 1, right);
        
        return mergeTwoLists(leftList, rightList);
    }
    
public:
    ListNode* mergeKLists(vector<ListNode*>& lists) {
        if (lists.empty()) return nullptr;
        return mergeKListsHelper(lists, 0, lists.size() - 1);
    }
};

/*
 * APPROACH 3: ITERATIVE PAIRWISE MERGE
 * 
 * Intuition:
 * - Merge lists pairwise iteratively
 * - Reduce k lists to k/2, then k/4, etc.
 * - Bottom-up approach
 * 
 * Algorithm:
 * 1. While more than one list:
 *    - Merge pairs of lists
 *    - Store merged lists
 * 2. Return final list
 * 
 * Time Complexity: O(N log k)
 * Space Complexity: O(1)
 */
class Solution3 {
private:
    ListNode* mergeTwoLists(ListNode* l1, ListNode* l2) {
        ListNode dummy(0);
        ListNode* tail = &dummy;
        
        while (l1 && l2) {
            if (l1->val < l2->val) {
                tail->next = l1;
                l1 = l1->next;
            } else {
                tail->next = l2;
                l2 = l2->next;
            }
            tail = tail->next;
        }
        
        tail->next = l1 ? l1 : l2;
        return dummy.next;
    }
    
public:
    ListNode* mergeKLists(vector<ListNode*>& lists) {
        if (lists.empty()) return nullptr;
        
        while (lists.size() > 1) {
            vector<ListNode*> merged;
            
            for (int i = 0; i < lists.size(); i += 2) {
                ListNode* l1 = lists[i];
                ListNode* l2 = (i + 1 < lists.size()) ? lists[i + 1] : nullptr;
                merged.push_back(mergeTwoLists(l1, l2));
            }
            
            lists = merged;
        }
        
        return lists[0];
    }
};

/*
 * APPROACH 4: SEQUENTIAL MERGE
 * 
 * Intuition:
 * - Merge lists one by one
 * - Merge first two, then merge result with third, etc.
 * - Simple but not optimal
 * 
 * Algorithm:
 * 1. Start with first list
 * 2. Merge with each subsequent list
 * 3. Return final merged list
 * 
 * Time Complexity: O(N * k) - less efficient
 * Space Complexity: O(1)
 */
class Solution4 {
private:
    ListNode* mergeTwoLists(ListNode* l1, ListNode* l2) {
        ListNode dummy(0);
        ListNode* tail = &dummy;
        
        while (l1 && l2) {
            if (l1->val < l2->val) {
                tail->next = l1;
                l1 = l1->next;
            } else {
                tail->next = l2;
                l2 = l2->next;
            }
            tail = tail->next;
        }
        
        tail->next = l1 ? l1 : l2;
        return dummy.next;
    }
    
public:
    ListNode* mergeKLists(vector<ListNode*>& lists) {
        if (lists.empty()) return nullptr;
        
        ListNode* result = lists[0];
        for (int i = 1; i < lists.size(); i++) {
            result = mergeTwoLists(result, lists[i]);
        }
        
        return result;
    }
};

/*
 * APPROACH 5: PRIORITY QUEUE WITH CUSTOM COMPARATOR
 * 
 * Intuition:
 * - Similar to Approach 1
 * - Use struct for cleaner comparator
 * - More explicit implementation
 * 
 * Algorithm:
 * Same as Approach 1 with struct comparator
 * 
 * Time Complexity: O(N log k)
 * Space Complexity: O(k)
 */
class Solution5 {
private:
    struct Compare {
        bool operator()(ListNode* a, ListNode* b) {
            return a->val > b->val;
        }
    };
    
public:
    ListNode* mergeKLists(vector<ListNode*>& lists) {
        priority_queue<ListNode*, vector<ListNode*>, Compare> pq;
        
        for (ListNode* list : lists) {
            if (list) {
                pq.push(list);
            }
        }
        
        ListNode dummy(0);
        ListNode* tail = &dummy;
        
        while (!pq.empty()) {
            ListNode* node = pq.top();
            pq.pop();
            
            tail->next = node;
            tail = tail->next;
            
            if (node->next) {
                pq.push(node->next);
            }
        }
        
        return dummy.next;
    }
};

// Helper functions
ListNode* createList(vector<int> vals) {
    ListNode dummy(0);
    ListNode* tail = &dummy;
    for (int val : vals) {
        tail->next = new ListNode(val);
        tail = tail->next;
    }
    return dummy.next;
}

void printList(ListNode* head) {
    cout << "[";
    while (head) {
        cout << head->val;
        if (head->next) cout << ",";
        head = head->next;
    }
    cout << "]";
}

// Test function
void test(vector<vector<int>> input, int approach) {
    vector<ListNode*> lists;
    for (auto& vals : input) {
        lists.push_back(createList(vals));
    }
    
    cout << "Input: [";
    for (int i = 0; i < input.size(); i++) {
        cout << "[";
        for (int j = 0; j < input[i].size(); j++) {
            cout << input[i][j];
            if (j < input[i].size() - 1) cout << ",";
        }
        cout << "]";
        if (i < input.size() - 1) cout << ",";
    }
    cout << "]\n";
    
    ListNode* result = nullptr;
    
    switch(approach) {
        case 1: {
            Solution1 sol;
            result = sol.mergeKLists(lists);
            cout << "Approach 1 (Min Heap): ";
            break;
        }
        case 2: {
            Solution2 sol;
            result = sol.mergeKLists(lists);
            cout << "Approach 2 (Divide & Conquer): ";
            break;
        }
        case 3: {
            Solution3 sol;
            result = sol.mergeKLists(lists);
            cout << "Approach 3 (Iterative Pairwise): ";
            break;
        }
        case 4: {
            Solution4 sol;
            result = sol.mergeKLists(lists);
            cout << "Approach 4 (Sequential): ";
            break;
        }
        case 5: {
            Solution5 sol;
            result = sol.mergeKLists(lists);
            cout << "Approach 5 (PQ Custom): ";
            break;
        }
    }
    
    printList(result);
    cout << "\n\n";
}

int main() {
    // Test Case 1: Standard case
    cout << "Test Case 1: Standard case\n";
    vector<vector<int>> test1 = {{1,4,5}, {1,3,4}, {2,6}};
    for (int i = 1; i <= 5; i++) {
        test(test1, i);
    }
    
    // Test Case 2: Empty lists
    cout << "Test Case 2: Empty\n";
    vector<vector<int>> test2 = {};
    for (int i = 1; i <= 5; i++) {
        test(test2, i);
    }
    
    // Test Case 3: Single list
    cout << "Test Case 3: Single list\n";
    vector<vector<int>> test3 = {{1,2,3}};
    for (int i = 1; i <= 5; i++) {
        test(test3, i);
    }
    
    return 0;
}

/*
 * COMPLEXITY ANALYSIS SUMMARY:
 * 
 * Approach 1 (Min Heap - OPTIMAL):
 * - Time: O(N log k) where N = total nodes
 * - Space: O(k) - heap size
 * - Best for: Standard solution, efficient
 * 
 * Approach 2 (Divide & Conquer):
 * - Time: O(N log k)
 * - Space: O(log k) - recursion
 * - Best for: Recursive preference
 * 
 * Approach 3 (Iterative Pairwise):
 * - Time: O(N log k)
 * - Space: O(1)
 * - Best for: Space efficiency
 * 
 * Approach 4 (Sequential):
 * - Time: O(N * k) - less efficient
 * - Space: O(1)
 * - Best for: Simple implementation
 * 
 * Approach 5 (PQ Custom):
 * - Time: O(N log k)
 * - Space: O(k)
 * - Best for: Clean comparator
 * 
 * INTERVIEW TIPS:
 * 1. Start with Approach 1 (min heap)
 * 2. Explain why heap is efficient
 * 3. Mention divide & conquer alternative
 * 4. Draw example with 3 lists
 * 5. Discuss time complexity clearly
 * 
 * COMMON MISTAKES:
 * 1. Not handling empty lists
 * 2. Wrong heap comparator
 * 3. Forgetting to add next node to heap
 * 4. Memory leaks in test code
 * 5. Off-by-one in pairwise merge
 * 
 * FOLLOW-UP QUESTIONS:
 * 1. What if lists are very long? (Heap still optimal)
 * 2. Can you do it in-place? (Yes, just rearrange pointers)
 * 3. What about k=2? (Simple two-pointer merge)
 * 4. How to handle duplicates? (Already handled)
 * 5. What if k is very large? (Heap still O(log k))
 * 
 * RELATED PROBLEMS:
 * - Merge Two Sorted Lists
 * - Merge Sorted Array
 * - Kth Smallest Element in Sorted Matrix
 * - Find K Pairs with Smallest Sums
 * - Ugly Number II
 * 
 * KEY INSIGHTS:
 * 1. Min heap tracks smallest among k lists
 * 2. Heap size never exceeds k
 * 3. Each node processed once
 * 4. Divide & conquer also O(N log k)
 * 5. Sequential merge is O(N * k)
 * 
 * WHY MIN HEAP WORKS:
 * - Always extract minimum element
 * - Heap maintains k candidates
 * - Each extraction is O(log k)
 * - Total N extractions
 * - Overall O(N log k)
 * 
 * DIVIDE & CONQUER ANALYSIS:
 * - Merge pairs: k/2 merges of 2N/k nodes
 * - Next level: k/4 merges of 4N/k nodes
 * - log k levels total
 * - Each level processes N nodes
 * - Total: O(N log k)
 * 
 * COMPARISON:
 * - Heap: O(N log k), O(k) space
 * - D&C: O(N log k), O(log k) space
 * - Iterative: O(N log k), O(1) space
 * - Sequential: O(N * k), O(1) space
 * - Choose based on constraints
 */

// Made with Bob
